### PR TITLE
WIP: Add gtk

### PR DESCRIPTION
**Pull request by jakirkham**
Fixes https://github.com/conda-forge/staged-recipes/issues/3914
Requires https://github.com/conda-forge/atk-feedstock/pull/1

Attempts to build GTK 2 from source on Linux and macOS. Feedback and other maintainers welcome.

### CODE REVIEW

File: recipes/gtk/build.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+
+./configure --prefix="${PREFIX}"
+
+make -j$CPU_COUNT
+# make check -j$CPU_COUNT
```
Comment (jakirkham) [inline recipes/gtk/build.sh]: Testing this is going to be a bit complex. Suggestions welcome.

File: recipes/gtk/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  url: http://ftp.gnome.org/pub/GNOME/sources/gtk+/{{ major_minor }}/gtk+-{{ version }}.tar.xz
+  sha256: 68c1922732c7efc08df4656a5366dcc3afdc8791513400dac276009b40954658
+
+build:
+  number: 0
+  skip: true  # [win]
+  # Requires Cairo's X11 libs.
+  # Please see issue ( https://github.com/conda-forge/cairo-feedstock/issues/22 ).
+  # Also please see PR ( https://github.com/conda-forge/cairo-feedstock/pull/23 ).
+  skip: true  # [osx]
```
Comment (jakirkham) [inline recipes/gtk/meta.yaml]: Getting macOS support will require Cairo having X11 libraries. Please see issue ( https://github.com/conda-forge/cairo-feedstock/issues/22 ) and PR ( https://github.com/conda-forge/cairo-feedstock/pull/23 ) for more details.

**Comment by jakirkham** [comment]
This is blocked by PR ( https://github.com/conda-forge/atk-feedstock/pull/1 ) ATM. Once that is merged and packages released, we can revisit this.

**Comment by jakirkham** [comment]
cc @ocefpaf @pkgw @sdvillal @Korijn

**Comment by jakirkham** [comment]
Built `atk` and `gtk` for Linux and have placed them in my channel, [`jakirkham`]( https://anaconda.org/jakirkham/ ), if anyone would like to test them.

**Comment by jakirkham** [comment]
cc @ccordoba12

**Comment by sdvillal** [comment]
I agree that testing can be a bit tricky. I have installed your packages, look at how DSOs get resolved and it all looks good to me. Perhaps what we should do is to build and run something against this to exercise it a bit.

[Here is our old recipe](https://github.com/loopbio/gtk2-feedstock) if it can be of help.

**Comment by ccordoba12** [comment]
Thanks for this @jakirkham! Unfortunately, I can offer a hand here because I've never compiled Gtk.

**Comment by jakirkham** [comment]
Borrowed some common flags used to build GTK from various sources (mainly Homebrew and Linuxbrew though MacPorts and the loopbio recipe were looked at as well). Have enabled the macOS build using Quartz (not X11), which is consistent with these sources and seems to build ok. Also made some minor tweaks to the tests at the end.

**Comment by jakirkham** [comment]
Have uploaded new packages to my own channel, [`jakirkham`]( https://anaconda.org/jakirkham/ ), for `gtk`. Overwrote the old packages in the process for simplicity.

File: recipes/gtk/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  number: 0
+  skip: true  # [win]
+  detect_binary_files_with_prefix: true
+
+requirements:
+  build:
+    - toolchain
+    - pkg-config
+    - perl 5.22.2.1
+    - atk
```
Comment (jakirkham) [inline recipes/gtk/meta.yaml]: Not sure how we want to pin this.

File: recipes/gtk/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    - cairo 1.14.*
+    - fontconfig 2.12.*
+    - freetype 2.7
+    - gdk-pixbuf
+    - gettext
```
Comment (jakirkham) [inline recipes/gtk/meta.yaml]: Not sure how we want to pin these.

File: recipes/gtk/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    - glib 2.51.*
+    - gobject-introspection
+    - graphite2
```
Comment (jakirkham) [inline recipes/gtk/meta.yaml]: Not sure how we want to pin these.

File: recipes/gtk/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    - harfbuzz 1.3.4
+    - icu 58.*
+    - jpeg 9*
+    - libffi
+    - libiconv
```
Comment (jakirkham) [inline recipes/gtk/meta.yaml]: Not sure how we want to pin these, if at all.

File: recipes/gtk/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    - libpng >=1.6.28,<1.7
+    - libtiff >=4.0.3,<4.0.8
+    - libxml2 2.9.*
+    - pango 1.40.*
+    - pcre
```
Comment (jakirkham) [inline recipes/gtk/meta.yaml]: Not sure how we want to pin this.

File: recipes/gtk/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+about:
+  home: https://developer.gnome.org/gtk/
+  license: LGPL-2
+  license_file: COPYING
+  summary: Primary library used to construct user interfaces in GNOME applications.
+
+extra:
+  recipe-maintainers:
+    - jakirkham
```
Comment (jakirkham) [inline recipes/gtk/meta.yaml]: Anyone else want to help maintain? I'm afraid I can't promise lots of time to this myself.

Comment (pkgw) [reply recipes/gtk/meta.yaml]: I can help.

**Comment by pkgw** [comment]
Can we call this package "gtk2"? Gtk+3 has been around for a looooong time at this point and should be considered the "normal" Gtk. They are parallel-installable and it should be possible for users to have both versions 2 and 3 installed at the same time.

**Comment by jakirkham** [comment]
So this is an interesting question. We have not tended to have SO named packages (with some exceptions like `pcre` and `pcre2`). For instance, `qt` is not an SO named package. My initial thought was to do the same with `gtk`. This would make it easier if people want to select GTK 2 based packages versus GTK 3. Though it does seem like Linux distros, Homebrew, and MacPorts do use the SO name with GTK. So this would be pretty consistent with what other packagers have done. All of this being said, I have no real strong feelings on it. Though would like to see some buy-in for how we name it. Perhaps @conda-forge/core and @conda-forge/staged-recipes could weigh in.

**Comment by pkgw** [comment]
I wouldn't view this through the lens of merely the soname. GTK+2 and GTK+3 are two related but distinct software systems, and they are both large and sophisticated toolkits with large user bases. Different applications will use either one or the other but there's no reason to limit the set of applications that can be installed based on which toolkit they happen to use — unless the two versions can't exist side-by-side. But the GTK+ developers put in a whole lot of effort to make parallel installation possible, so we should take advantage of that.

**Comment by jakirkham** [comment]
Understood. The only point I was trying to raise was if a user wanted some package `foo` that could be built with GTK 2 or 3, how would they select this assuming there is a need. If we just have `gtk` without a version number, this is straightforward. If we embed the major version in the package name, we may need to give this some more thought. Outside of this issue, have no real strong feelings as already noted.

**Comment by scopatz** [comment]
It seems to me that for these packages the major version number should be embedded in the name. Maybe then gtk can be a meta-package that simply points to one or the other.

**Comment by pkgw** [comment]
That's a choice to be made by the packager, not the user, no? If I'm packaging the `foo` application I can choose which one it will depend on. I don’t see there being much reason to try to provide both alternatives to users.

**Comment by jakirkham** [comment]
> That's a choice to be made by the packager, not the user, no?

Sure. I overloaded user here. Meant a packager downstream of `gtk`. 😄

> If I'm packaging the `foo` application I can choose which one it will depend on. I don’t see there being much reason to try to provide both alternatives to users.

Yes, you can totally choose. Though the question is what if you want to support both `gtk` 2 and 3? How should this downstream packager handle such a case? Just want to make sure there is a standard framework for downstream packagers to handle such a problem.

**Comment by pkgw** [comment]
@jakirkham Based on a lot of experience with Gtk software packages over many years, I will assert that downstream packages virtually *never* have simultaneous support for Gtk2 and Gtk3; they either have one or the other. I am pretty sure that I've literally never seen this. That might be a bit less true for programs based on PyGtk since it's easier to be loosey-goosey with Python code than C, but even in that case, the recommended Python APIs changed substantially so one generally needed to make a firm choice. ([Personal experience here](https://github.com/pkgw/pwkit/commit/0570fc6b84407f42655308efb2546577c2882cf9)).

So basically I don't think issue of downstream packages that want to support *either* 2 or 3 is one that will arise in practice.

**Comment by jakirkham** [comment]
OpenCV can support Gtk2 or Gtk3.

**Comment by jakirkham** [comment]
cc @patricksnape

**Comment by sdvillal** [comment]
The goal being to have a consistent software stack, shouldn't the rule simply be like this? 
  - if two packages can be installed together, let them be installed together
 -  if two packages should not be installed together, do not let them be installed together 

Now, in the very simplified conda packaging world, where there are very few ways to express relationships between packages, we can read "let two packages be installed together" as "let two packages be named differently".

I think there should be no discussion about this. Installing gtk2 and gtk3 on the same environment should be possible. Therefore both packages should be named differently.

There is then no trouble downstream. For example, in the rare case we would want to have support for all three gtk2, gtk3 and qt4 as opencv highgui backends, we would then just have three versions of the opencv package bringing along their corresponding dependencies. The concrete version would be selected with the build string, or via a conda feature if you want. The user will install whatever version she needs. And that would not mean other opencv-unrelated packages depending on other gui toolkits to be affected.

The gtk2 + gtk3 should be allowed to be installed in the same environment is in clear contrast with other packages that should not be allowed to end up in the same environment but still are not named the same (thinking on the cpu vs gpu versions of some packages in defaults that can be happily and wrongly installed side by side).

**Comment by pkgw** [comment]
@jakirkham I stand corrected.

**Comment by jakirkham** [comment]
I'm sure you are right that this is hardly the norm. OpenCV is really a behmoth.

**Comment by jakirkham** [comment]
cc @oinizan

**Comment by jakirkham** [comment]
Going to try to bring this back to life.

> I think there should be no discussion about this. Installing gtk2 and gtk3 on the same environment should be possible. Therefore both packages should be named differently.

I don't think we are discussing this point any more. We are in agreement.

> There is then no trouble downstream. For example, in the rare case we would want to have support for all three gtk2, gtk3 and qt4 as opencv highgui backends, we would then just have three versions of the opencv package bringing along their corresponding dependencies. The concrete version would be selected with the build string, or via a conda feature if you want.

This is where we are having a disagreement I think. There is a challenge for downstream packagers to specify this in a consistent manner as there currently isn't one solution. 😄 So what I'm asking is for us to settle on one of these solutions. For BLAS selection, we have the `blas` package (feature-based), which could be an option here. Some have used build string style selection with the LLVM toolchain to some success. The feature solution is good if we want to force one GUI backend for all things. The build string solution is good if we want to allow different packages to use different GUI backends. Personally I don't have strong opinions as far as this goes with GUI backends (maybe build string?), but I do think it is important for us to agree on the solution and document it. Otherwise different packages may adopt different solutions, which would be a nightmare for end users. Picking a solution for the community to follow avoids that. Does this sound reasonable? Any other thoughts? 😉

**Comment by scopatz** [comment]
I personally find using features a little annoying to develop with.  If the string method works, I would be in favor of that.

**Comment by sdvillal** [comment]
I think that the only proper solution we currently have is to recommend the build string method. Hypothetical downstream dependents that provide two variants, one for gtk2 and one for gtk3, would be selected like ```downstream=3.3=*gtk2*```.

To my mind, conda features are not only annoying to use. They simply do not cut it here. Because gtk2 and gtk3 can be installed side by side, each of them need to track a different feature. This is true even with new conda 4.4 key-value features: the feature cannot be "gtk=[2|3]" (which would allow only one of the two gtk variants in the environment) but instead "gtk2=present" and "gtk3=present". If each of the two downstream package variants only differ on requiring one of two different features, how will the downstream package be resolved when both features are provided by the environment? I think randomly in conda < 4.4. In conda >= 4.4, if I understand correctly, we could maybe use the [features="gtk2=present"] feature, which seems clumsy.

To allow the user to choose one version or the other via conda features, we would also add a "downstream-gtk2-feature" and "downstream-gtk3-feature" per downstream package. I think that is painful too.

Having different downstream package names should also be discarded - for the obvious reason they won't be anymore variants of the same package but two different packages that could be installed on parallel and that would force further differentiation/split of dependents.

**Comment by sdvillal** [comment]
Just a side note: I think the "blas package tracking feature" is an unfortunate historical event. For example, openblas and MKL can, at the moment, both be installed on parallel. So this discussion could apply also to them.

**Comment by jakirkham** [comment]
Could you please weigh in on this one if you have time, @mingwandroid? It would be very helpful to hear your thoughts on this. In particular how we solve the Gtk2/Gtk3 issue going forward.

**Comment by mingwandroid** [comment]
I would ignore `gtk2` here. We should be thinking about `gtk4` now-a-days. I know I'm sidestepping the core issue, but I think we can reasonably consider `gtk2` as useful as `Qt4` and there's no way on earth I'd consider spending a minute's effort on `Qt4` (yeah we already have it, but that's due to historical efforts).

Anyway, if it is true that:

> gtk2 and gtk3 can be installed side by side, each of them need to track a different feature

.. then I don't agree at all with with:

> Having different downstream package names should also be discarded - for the obvious reason they won't be anymore variants of the same package but two different packages that could wrongly be installed on parallel and that would force further differentiation/split of dependents

Because I do not see it as painful at all for anyone. Legacy software that doesn't work with `gtk3` (or 4!) would be built against `gtk2` (or 3!) and everything else would be built against `gtk3` (or 4!) and they can both be installed in the same environment and I think that's completely fine personally.

**Comment by mingwandroid** [comment]
.. overall, the point I am making here is that there's no intrinsic 'value' in ever using `gtk2` when you have the option to use `gtk3`. There are no benefits conferred to the end user and they will never say "but I want the gtk2 variant" the same as they'll never say "I want Qt4 instead of Qt5".

**Comment by jakirkham** [comment]
As we do package legacy software in other cases where we need it, this doesn't seem that unusual. Given we are often using `gtk2` from `yum`, having this as a `conda` package would get us away from using `yum` to install it, which is already an improvement IMHO. Given this, are there any blockers to merging?

**Comment by jakirkham** [comment]
cc @conda-forge/staged-recipes

**Comment by scopatz** [comment]
I think we should take the plunge and fix things in the feedstock, as needed.
